internal/storage/chunk_uploader.go

```diff
@@ -0,0 +1,214 @@
+	// upload call times out.
+	StorageCopyTimeoutDuration time.Duration = time.Second
+)
+
+// A chunkUploader is an implementation of ChunkUploader
+// interface, which uses storage.Writer from go-storage-client
+// for resumable upload.
+//
+// It also stores the current state of the uploader.
+type chunkUploader struct {
+	// Internal objects for business logic.
+	wc         *storage.Writer
+	objectName string
+	mu         sync.Mutex
+
+	// Attributes for providing updates to user.
+	totalWriteInitiatedSoFar int64
+	// Total number of bytes successfully written by this writer so far.
+	// This should be used only through the method BytesUploadedSoFar for
+	// thread-safety.
+	totalWriteSucceededSoFar int64
+	userProgressFunc         func(int64)
+
+	// Internal state for lifecycle management.
+	state UploaderState
+}
+
+// NewChunkUploader creates a new instance of chunkUploader,
+// for the given inputs.
+func NewChunkUploader(ctx context.Context, obj *storage.ObjectHandle, req *gcs.CreateObjectRequest, writeChunkSize int, progressFunc func(int64)) (gcs.ChunkUploader, error) {
+	if ctx == nil {
+		return nil, fmt.Errorf("ctx is nil")
+	}
+	if obj == nil || req == nil {
+		return nil, fmt.Errorf("nil ObjectHandle or CreateObjectRequest")
+	}
+
+	if obj.ObjectName() != req.Name {
+		return nil, fmt.Errorf("names of passed ObjectHandle and CreateObjectRequest don't match: object-handle=%s createObjectRequest.Name=%s", obj.ObjectName(), req.Name)
+	}
+
+	if req.GenerationPrecondition != nil && *req.GenerationPrecondition != 0 {
+		return nil, fmt.Errorf("request received for pre-existing object %s, supported only for new objects", req.Name)
+	}
+
+	if writeChunkSize <= 0 {
+		return nil, fmt.Errorf("chunkSize <= 0")
+	}
+
+	// Raw initialization.
+	uploader := chunkUploader{}
+
+	// Store references to necessary parameters.
+	uploader.objectName = obj.ObjectName()
+	uploader.userProgressFunc = progressFunc
+
+	// Create a NewWriter with the requested attributes, using Go Storage Client.
+	// NewWriter never returns nil, so no nil-check is needed on it.
+	wc := obj.NewWriter(ctx)
+	wc = storageutil.SetAttrsInWriter(wc, req)
+	wc.ChunkSize = writeChunkSize
+	wc.ProgressFunc = func(n int64) {
+		atomic.StoreInt64(&uploader.totalWriteSucceededSoFar, n)
+		logger.Debugf("%d bytes copied so far for object/file %s. chunk-size = %d", n, req.Name, wc.ChunkSize)
+
+		if uploader.userProgressFunc != nil {
+			uploader.userProgressFunc(n)
+		}
+	}
+
+	uploader.wc = wc
+	uploader.state = Initialized
+
+	return &uploader, nil
+}
+
+// BytesUploadedSoFar returns the total number of bytes successfully uploaded
+// so far using this uploader.
+// This is thread-safe against the
+// in-progress calls to BytesUploadedSoFar/UploadChunkAsync/Close
+// invoked from other threads/go-routines.
+func (uploader *chunkUploader) BytesUploadedSoFar() int64 {
+	return atomic.LoadInt64(&uploader.totalWriteSucceededSoFar)
```

Comment:
`atomic.LoadInt64` is basically for thread-safety. It's needed because this variable is updated in the callback (different goroutine) and is read in the goroutine (calling the `uploader.Close()` or `uploader.Upload...`).

I didn't know that there was a clearner alternative to it. I'll switch to `atomic.Int64.Load()` .